src/sentry/hybridcloud/models/outbox.py

```diff
@@ -224,60 +227,206 @@ def process_coalesced(
+                assert (
+                    first_coalesced
+                ), "first_coalesced incorrectly set for non-empty coalesce group"
+
+                # Record the net queue time metric using the date the first message was added.
+                metrics.timing(
+                    "outbox.coalesced_net_queue_time",
+                    datetime.datetime.now(tz=datetime.UTC).timestamp()
+                    - first_coalesced.date_added.timestamp(),
+                    tags=tags,
+                )
+
+                # Store the representative (highest id) and timing values for later use.
+                coalesced_id = coalesced.id
+                # 'scheduled_from' is used for metrics and can serve as the original scheduled timestamp.
+                scheduled_from = first_coalesced.scheduled_from
+                date_added = first_coalesced.date_added
+
+                # Get a list of IDs for all messages in the group that are older than the coalesced record.
+                # This is our snapshot of messages to process (delete) in this coalesced group.
+                all_ids = list(
+                    self.select_coalesced_messages()
+                    .filter(id__lt=coalesced_id)
+                    .values_list("id", flat=True)
+                    .order_by("id")
+                )
+
+                # Reserve the messages for processing. By updating scheduled_for to a point
+                # in the future (now + 1 hour), we effectively signal that these messages are
+                # being processed, thereby preventing any other drain process from picking them up.
+                self.objects.filter(id__in=all_ids + [coalesced_id]).update(
+                    scheduled_for=timezone.now() + datetime.timedelta(hours=1)
+                )
+
+            # End of the selection/reservation phase. The transaction is now committed, and all locks
+            # are released. Yield the representative coalesced message so that remote processing (like
+            # sending signals) can be done without an open transaction.
+            try:
+                yield coalesced
+            except Exception:
+                # An exception occurred during processing (e.g. send_signal() raised an exception).
+                # Revert the scheduled_for values back to their original value (scheduled_from)
+                # so that the messages are eligible for reprocessing.
+                self.objects.filter(id__in=all_ids + [coalesced_id]).update(
+                    scheduled_for=scheduled_from
+                )
+                metrics.incr("outbox.coalesced_yield_error", tags=tags)
+                raise
+
+            # After the yield, we now proceed to clean up (delete) the reserved messages.
+            try:
+                deleted_count = 0
+                batch_size = 50  # Process deletions in batches to avoid long-running transactions.
```

Comment:
We might be able to delete in larger slices again, as we're not in the select_for_update transaction, and deleting 50-100 rows should be fast enough to not get stomped.